### ISSUE TITLE
Research: Learn from Clippy's architecture patterns for advanced linting features

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#62]
# Research: Learn from Clippy's Architecture Patterns

## Background

Clippy represents one of the most successful and mature linting architectures in the Rust ecosystem. As we prepare for Phase 3 (LSP implementation) and future scaling of mdbook-lint, we should examine Clippy's proven patterns to identify improvements we could adopt.

**Note**: Our current architecture already implements single-pass AST parsing via `check_document_optimized()` - we parse once and share the AST with all rules. This is good\! But there are other patterns we could learn from.

### 1. 🏗️ Structured Lint Declaration

**Clippy's Approach:**
```rust
declare_clippy_lint\! {
    /// ### What it does
    /// Checks for `#[macro_use] use...`.
    /// ### Why is this bad?
    /// Since the Rust 2018 edition...
    /// ### Example
    /// ```rust,ignore
    /// #[macro_use]
    /// extern crate some_crate;
    /// ```
    #[clippy::version = "1.44.0"]
    pub MACRO_USE_IMPORTS,
    pedantic,
    "#[macro_use] is no longer needed"
}
```

**For mdbook-lint:**
```rust
declare_mdbook_lint\! {
    /// ### What it does  
    /// Checks for headings that skip hierarchy levels (e.g., H1 -> H3)
    /// ### Why is this bad?
    /// Skipping levels breaks document structure and accessibility
    /// ### Example
    /// ```markdown
    /// # Level 1
    /// ### Level 3 - BAD, skipped level 2
    /// ```
    #[mdbook_lint::version = "0.3.0"]
    pub MD001,
    Structure,
    "heading levels should increment by one level at a time"
}
```

**Benefits:**
- Consistent metadata across all rules
- Structured documentation generation
- Version tracking for rule introductions
- Better tooling integration

### 2. 📊 Enhanced Category System

**Clippy's Categories:**
- `correctness` - Code that is outright wrong
- `suspicious` - Code that is most likely wrong  
- `style` - Code that should be more idiomatic
- `complexity` - Complex code that could be simpler
- `perf` - Code that could run faster
- `pedantic` - Strict lints with possible false positives
- `restriction` - Prevent certain language features
- `nursery` - New/experimental lints
- `cargo` - Cargo manifest lints

**Our Current Categories:**
- Structure, Formatting, Links, Whitespace, mdBook

**Proposed Enhanced Categories:**
- `Correctness` - Objectively wrong markdown (broken links, malformed syntax)
- `Structure` - Document hierarchy and organization
- `Style` - Idiomatic markdown patterns
- `Accessibility` - Screen reader and accessibility concerns  
- `Performance` - Large images, excessive nesting affecting render speed
- `Pedantic` - Strict style rules that might have false positives
- `Compatibility` - Cross-platform and tool compatibility
- `mdBook` - mdBook-specific features and requirements

### 3. 🚀 Two-Pass Architecture for LSP

**Clippy's Approach:**
- `EarlyLintPass` - Fast AST-only checks
- `LateLintPass` - Slower type-aware checks

**For mdbook-lint LSP:**
```rust
// Fast pass for real-time feedback (as-you-type)
pub trait FastMarkdownPass {
    fn check_syntax(&self, document: &Document) -> Result<Vec<Violation>>;
}

// Semantic pass for comprehensive analysis (on save/request)  
pub trait SemanticMarkdownPass {
    fn check_semantics(&self, document: &Document, context: &ProjectContext) -> Result<Vec<Violation>>;
}
```

**Use Cases:**
- **Fast Pass**: Syntax errors, basic formatting, heading structure
- **Semantic Pass**: Link validation, cross-references, image optimization, spelling

### 4. ⚙️ Advanced Configuration System

**Clippy's `clippy.toml`:**
```toml
cognitive-complexity-threshold = 30
doc-valid-idents = ["GitHub", "OAuth"]
disallowed-names = ["foo", "baz"]
```

**Enhanced `.mdbook-lint.toml`:**
```toml
[rules.MD013]
line-length = 100
ignore-code-blocks = true
ignore-tables = true

[rules.MD045] 
allowed-extensions = [".png", ".jpg", ".svg"]
max-image-size-kb = 500

[accessibility]
require-alt-text = true
heading-anchor-style = "github"

[performance] 
max-image-width = 1200
warn-large-files-kb = 100
```

### 5. 🛠️ Utility Infrastructure  

**Clippy's `clippy_utils`:**
- Common AST traversal patterns
- Cached path resolution
- Diagnostic helpers
- Type checking utilities

**Proposed `mdbook_lint_utils`:**
```rust
pub mod ast_utils {
    pub fn find_all_headings(ast: &AstNode) -> Vec<HeadingInfo>;
    pub fn extract_links(ast: &AstNode) -> Vec<LinkInfo>; 
    pub fn find_code_blocks(ast: &AstNode) -> Vec<CodeBlockInfo>;
}

pub mod link_utils {
    pub struct LinkResolver {
        cache: HashMap<String, LinkStatus>,
    }
    
    impl LinkResolver {
        pub async fn resolve_link(&mut self, url: &str) -> LinkStatus;
        pub fn resolve_internal_link(&self, path: &str, base: &Path) -> Result<PathBuf>;
    }
}

pub mod diagnostics {
    pub fn span_lint_with_suggestion(
        rule: &dyn Rule,
        span: TextRange,
        message: &str,
        suggestion: Option<&str>
    ) -> Violation;
}
```

### 6. 📈 Performance Optimizations

**Clippy's Optimizations:**
- Rule filtering (only run enabled rules)
- Lazy evaluation with `OnceLock`
- MSRV-based filtering
- Path resolution caching

**For mdbook-lint:**
- **Rule Filtering**: Already implemented ✅
- **Lazy Link Resolution**: Cache HTTP requests and file checks
- **Incremental Linting**: Only re-lint changed sections (LSP)
- **Parallel Rule Execution**: Run independent rules concurrently

### Phase 1: Enhanced Metadata (Post-LSP)
- Implement `declare_mdbook_lint\!` macro
- Add version tracking to rules
- Enhance rule documentation structure

### Phase 2: Two-Pass Architecture (LSP Integration)
- Separate fast/semantic passes for LSP
- Implement incremental linting
- Add performance monitoring

### Phase 3: Advanced Configuration  
- Typed configuration with validation
- Rule-specific configuration
- Environment variable support

### Phase 4: Utility Infrastructure
- Extract `mdbook_lint_utils` crate
- Implement cached link resolution
- Add diagnostic helpers

## Questions for Discussion

1. **Priority**: Which patterns would provide the most immediate value?
2. **LSP Integration**: Should we implement two-pass architecture as part of LSP development?
3. **Breaking Changes**: How do we balance improvements vs. backward compatibility?
4. **Performance**: What are our current performance bottlenecks to prioritize?

## Examples for Our Rules

Here's how some of our existing rules could benefit:

**MD001 (Heading Increment):**
```rust
declare_mdbook_lint\! {
    /// ### What it does
    /// Checks for headings that skip hierarchy levels
    /// ### Why is this bad?  
    /// Breaks document structure and accessibility tools
    /// ### Example
    /// ```markdown
    /// # Chapter 1
    /// ### Section 1.1 ❌ (skipped H2)
    /// ## Section 1.1 ✅ (correct)
    /// ```
    #[mdbook_lint::version = "0.1.0"]
    pub MD001,
    Structure,
    "heading levels should increment by one level at a time"
}
```

**MD013 (Line Length):**
```rust
declare_mdbook_lint\! {
    /// ### What it does
    /// Checks for lines longer than configured maximum
    /// ### Why is this bad?
    /// Long lines reduce readability and cause horizontal scrolling
    /// ### Configuration
    /// ```toml
    /// [rules.MD013]
    /// line-length = 80
    /// ignore-code-blocks = true
    /// ```
    #[mdbook_lint::version = "0.1.0"]
    pub MD013,
    Style,
    "line length should not exceed configured maximum"
}
```

This research provides a roadmap for scaling mdbook-lint's architecture using proven patterns from one of Rust's most successful linting tools.

**Comment by joshrotenberg** [comment]
This research task supports the Rule Enhancement Roadmap #158 and Documentation Roadmap #159. Clippy's architecture patterns could inform our plugin system and rule development approach.
